sys/kern/vm_map.c

```diff
@@ -161,33 +165,49 @@ void vm_map_entry_destroy(vm_map_t *map, vm_map_entry_t *ent) {
   vm_map_entry_free(ent);
 }
 
+static inline vm_map_entry_t *vm_map_entry_copy(vm_map_entry_t *src) {
+  vm_object_hold(src->object);
+  vm_map_entry_t *new = vm_map_entry_alloc(src->object, src->start, src->end,
+                                           src->prot, src->flags);
+  return new;
+}
+
+/* Split vm_map_entry into two not empty entries.
+ * Returns entry which is after base entry. */
+static vm_map_entry_t *vm_map_entry_split(vm_map_t *map, vm_map_entry_t *ent,
+                                          vaddr_t splitat) {
+  assert(mtx_owned(&map->mtx));
+  assert(ent->start < splitat && splitat + 1 < ent->end);
```

Comment:
Why `splitat + 1 < ent->end` not `splitat < ent->end`?

Reply:
Hmm, it should be even stronger. We don't want to have empty `vm_map_entries` and smallest `vm_map_entry` describes at leas one page of memory. `splitat` should be page aligned and smaller than `ent->end`. I will fix that.